nemo/collections/asr/parts/submodules/jasper.py

```diff
@@ -958,3 +958,80 @@ def forward(self, input_: Tuple[List[Tensor], Optional[Tensor]]):
             return xs + [out], lens
 
         return [out], lens
+
+
+class ParallelBlock(nn.Module):
+    """
+    Computational module that computes several `blocks` independently from each other and aggregates the outputs.
+    Due to a particular way of structuring inputs with sequence masks in Jasper-like blocks, this module focuses
+    specifically on the Jasper family.
+
+    Args:
+        blocks: List of Jasper blocks that will be computed concurently. It is expected that they accept the same
+            input and return outputs with the same number of channels.
+        aggregation_mode: an optional string, indicating how the outputs will be aggregated. Supported values are
+            ['sum', 'dropout']. "sum" value forces outputs to be summed together. "dropout" value enables tower
+            dropout training with different blocks being dropped out during training.
+        block_dropout_prob: a probability of dropping any individual block during training with "dropout" aggregation
+            mode. Acts as a regularization technique.
+    """
+
+    def __init__(self, blocks, aggregation_mode: str = "sum", block_dropout_prob: int = 0.0):
+        super().__init__()
+        self.blocks = nn.ModuleList(blocks)
+        self.aggregation_mode = aggregation_mode
+        if aggregation_mode == "dropout":
+            self.weights = nn.Parameter(torch.ones(1, len(blocks), 1, 1), requires_grad=False)
+            self.dropout = nn.Dropout(block_dropout_prob)
+
+    def get_dropout_mask(self):
+        weights = self.dropout(self.weights)
+        while torch.sum(weights) == 0 and self.dropout.p < 1.0:
+            weights = self.dropout(self.weights)
+        return weights
+
+    def forward(self, x: Tuple[List[Tensor], Optional[Tensor]]):
+        """
+        Forward pass computing aggregated output.
+
+        Args:
+            x: tuple of padded signal and lengths the signal. The shape of the signal is [B, D, T]. The lengths are
+                1D torch tensor of length B.
+
+        Returns:
+           torch tensor after passing input throught each block and aggregating these outputs according to the
+           aggregation mode.
+        """
+        if len(self.blocks) == 1:
+            return self.blocks[0](x)
+
+        input_feat = x[0][-1]
+        in_channels = input_feat.shape[-2]
+
+        result = None
+        max_mask = None
+
+        scaling_weights = None
+        if self.aggregation_mode == "dropout":
+            scaling_weights = self.get_dropout_mask()
+
+        for i, block in enumerate(self.blocks):
+            output, mask = block(x)
+
+            weighted_output = output[-1]
+            out_channels = weighted_output.shape[-2]
+            if self.aggregation_mode == "dropout":
+                weighted_output = scaling_weights[:, i, :, :] * output[-1]
```

Comment:
This is just pointwise scaling, is it necessary to use a rank 4 tensor to store this weights?

Reply:
This was legacy from the time where we also had per-element scaling. 

I changed it to 1D tensor. This will break some of the previous tower dropout checkpoints but that's not an issue right now, since we're not actively using them.